policies/templates/gcp_restricted_firewall_rules_v1.yaml

```diff
@@ -0,0 +1,627 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# This template is for policies restricting which firewall rules
+# should exist in a specific location in GCP.
+# All firewall rules matching the passed parameters will be flagged as violations.
+# All parameters are optional.
+#
+# A firewall rule will trigger a violation ONLY IF it matches ALL the passed parameters.
+
+apiVersion: templates.gatekeeper.sh/v1alpha1
+kind: ConstraintTemplate
+metadata:
+  name: gcp-restricted-firewall-rules-v1
+spec:
+  crd:
+    spec:
+      names:
+        kind: GCPRestrictedFirewallRulesConstraintV1
+        plural: gcprestrictedfirewallrulesconstraintsv1
+      validation:
+        openAPIV3Schema:
+          properties:
+            rules:
+              description: "List of rules to match in order to raise violations. Rule are object describing rules (see below)"
+              type: array
+              items:
+                type: objects
+                properties:
+                  direction:
+                    description: "Direction of the rules to restrict."
+                    type: string
+                    enum: ["INGRESS", "EGRESS", "any"]
+                    default: "any"
+                  rule_type:
+                    description: "Type of the rules to restrict."
+                    type: string
+                    enum: ["allowed", "denied", "any"]
```

Comment:
Shouldn't this be `allow` and `deny` to match the standard language?

Reply:
Well I matched the CAI terminology here. I could update this to allow or deny but the test for this field would be slightly redundant. Instead of one simple:

```
# test that the key exists in the fw_rule (allowed or denied)
        	fw_rule[lower(rule_type)]
```

we would need to split this into **2 functions** which would look like:

```
rule_type == "allow"
# test that the key exists in the fw_rule
        	fw_rule["allowed"]
```

Is this ok with you?

Reply:
If that's the CAI terminology, I'm fine with keeping it as-is.